### PR TITLE
feat: code field in form

### LINKED ISSUE #174: feat: add multiline support for input
## Description
It would be nice if the "Workflow Parameters" and "Workflow Engine Parameters" would also accept YAML (which would then be converted to JSON), as it is easier for humans to manipulate YAML. For example, {"input":{"class":"File","path":"https://is.muni.cz/pics/design/r/znak_MU.png"}} is much harder to type/construct than:
```
input:
  class: File
  path: https://is.muni.cz/pics/design/r/znak_MU.png
```
## Possible solution(s)
Maybe we support multiline inputs to the form so that `package-author` can specifically parse that input from YAML to JSON.

### PR DESCRIPTION
**Pull request by JaeAeich**
## Description

Add code editor, to form, this will help users take input of schema like executors etc in `YAML`, `JSON` format.  

Fixes #174 

## Checklist

- [x] My code follows the [contributing guidelines][contributing] of this project.
- [x] I am aware that all my commits will be squashed into a single commit, using the PR title as the commit message.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code in hard-to-understand areas.
- [x] I have updated the user-facing documentation to describe any new or changed behavior.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have not reduced the existing code coverage.

## Comments

[contributing]: CONTRIBUTING.md

### CODE REVIEW

**Comment by JaeAeich** [comment]
@anuragxxd @SalihuDickson can you guys please help me figure why the css parts are not changing the label style. Checkout the styles component in the docs.

PS: I've added CSS parts and I think this might be due to a left out CSS part or some irregularity in previously existing parts.

**Comment by SalihuDickson** [comment]
hey @JaeAeich, you can't change the style because you haven't added "label" as one of the css parts for the code component. You need to go into the code component and add a part for the label for that component, you can then export that css part as "label" in the form component.

**Comment by JaeAeich** [comment]
@SalihuDickson still same, plus if you look at it the styles are getting applied in places they shouldn't be, like the border which includes the text field as well as label. Lemme know if you find something, Illl look into it as well :)

File: packages/ecc-utils-design/src/components/form/form.ts [hunk 1]
```diff
@@ -246,6 +257,34 @@ export default class EccUtilsDesignForm extends LitElement {
...
 
+  private renderCodeTemplate(field: Field, path: string): TemplateResult {
+    if (field.type !== "code") return html``;
+    // TODO: add CSS parts
+
+    const { fieldOptions, codeOptions } = field;
+    const { formControl, formControlLabel } = this.cssParts;
+
+    return html`<ecc-utils-design-code
+      part="form-control: ${formControl},
```
Comment (SalihuDickson) [inline packages/ecc-utils-design/src/components/form/form.ts:268]: by setting the part like this you're setting these parts to control the entire component and that messing with its styles. You only need to export the already created parts with `exportparts`.

File: packages/ecc-utils-design/src/components/form/form.ts [hunk 2]
```diff
@@ -246,6 +257,34 @@ export default class EccUtilsDesignForm extends LitElement {
...
+			form-control-label: ${formControlLabel}, 
+			input: ${this.cssParts.input},
+			base: ${this.cssParts.inputBase}
+      label: ${this.cssParts.label}"
+      exportparts="form-control: ${formControl},
+			form-control-label: ${formControlLabel}, 
+			input: ${this.cssParts.input}, 
+			base: ${this.cssParts.inputBase}
```
Comment (SalihuDickson) [inline packages/ecc-utils-design/src/components/form/form.ts:276]: finally, you're missing a comma at the end of this line.

**Review by SalihuDickson — CHANGES_REQUESTED** [review]
Correcting these should fix your issue. Since these are minor changes i didn't want to go through the trouble of creating a PR for them but if that's what you'd prefer please let me know.

File: packages/ecc-utils-design/src/components/form/form.ts (hunk 1 above)
Comment (JaeAeich) [reply packages/ecc-utils-design/src/components/form/form.ts]: please check if this is what you meant.

File: apps/documentation/components/packages/design/form/complexForm.tsx [hunk 3]
```diff
@@ -37,6 +37,20 @@ export default function Form() {
         tooltip: 'Your email address',
       },
     },
+    {
```
Comment (SalihuDickson) [inline apps/documentation/components/packages/design/form/complexForm.tsx:40]: add this code component to field object below that displays the code people can copy to get the data.

File: apps/documentation/components/packages/design/form/complexForm.tsx [hunk 4]
```diff
@@ -74,6 +88,20 @@ export default function Form() {
                 tooltip: 'Your house number',
               },
             },
+            {
```
Comment (SalihuDickson) [inline apps/documentation/components/packages/design/form/complexForm.tsx:91]: add this code component to field object below that displays the code people can copy to get the data.

File: apps/documentation/pages/design/components/form.mdx [hunk 5]
```diff
@@ -110,7 +110,9 @@ This property is used to render the fields in the form. Fields can be passed as
...
 |           children            | `false`  | `null`  | `array`                                                                                                                             | Children fields for the field if type is array. This allows fields to be added dynamically                                                                                                                                |
+|          codeOptions          | `false`  | `null`  | `{indentation, language, blurDelay}`                                                                                                | Used to configure code editor fields, please look for docs [here](./code#properties).                                                                                                                                     |
```
Comment (SalihuDickson) [inline apps/documentation/pages/design/components/form.mdx:113]: please list out the nested options and their description like was done for the fieldOptions

File: packages/ecc-utils-design/demo/form/index.html
Comment (SalihuDickson) [inline packages/ecc-utils-design/demo/form/index.html:1]: please revert the formatting, done on this page.

File: packages/ecc-utils-design/src/components/form/form.ts [hunk 6]
```diff
@@ -438,7 +475,9 @@ export default class EccUtilsDesignForm extends LitElement {
     if (field.type === "array") {
       return this.renderArrayTemplate(field, newPath);
     }
-
+    if (field.type === "code") {
```
Comment (SalihuDickson) [inline packages/ecc-utils-design/src/components/form/form.ts:478]: this should be below the if statement that checks if the field is required but empty.

**Review by SalihuDickson — CHANGES_REQUESTED** [review]

File: apps/documentation/components/packages/design/form/complexForm.tsx (hunk 4 above)
Comment (JaeAeich) [reply apps/documentation/components/packages/design/form/complexForm.tsx:91]: Sorry, I don't understand!

**Comment by anuragxxd** [comment]
Check out the #324 have replaced the code components there with ace editor.

**Comment by anuragxxd** [comment]
Hey @JaeAeich, are you working on this PR or can we close this?

**Pull request closed without merging**